src/crypto-providers/signingValidation.ts

```diff
@@ -0,0 +1,232 @@
+  let numStakeWitnesses = unsignedTxParsed.withdrawals.length
+  let numPoolColdWitnesses = 0
+  unsignedTxParsed.certificates.forEach((cert) => {
+    switch (cert.type) {
+      case TxCertificateKeys.STAKING_KEY_REGISTRATION:
+      case TxCertificateKeys.STAKING_KEY_DEREGISTRATION:
+      case TxCertificateKeys.DELEGATION:
+        numStakeWitnesses += 1
+        break
+
+      case TxCertificateKeys.STAKEPOOL_RETIREMENT:
+        numPoolColdWitnesses += 1
+        break
+
+      default:
+        break
+    }
+  })
+  return { numStakeWitnesses, numPoolColdWitnesses }
+}
+
+const validateOrdinaryTx = (unsignedTxParsed: _UnsignedTxParsed) => {
+  _validateNonMultisigTx(unsignedTxParsed)
+
+  if (unsignedTxParsed.inputs.length === 0) {
+    throw Error(Errors.MissingInputError)
+  }
+}
+
+// TODO we don't count unique witnesses
+const validateOrdinaryWitnesses = (params: SigningParameters) => {
+  const {
+    stakeSigningFiles, poolColdSigningFiles, mintSigningFiles, multisigSigningFiles,
+  } = filterSigningFiles(params.hwSigningFileData)
+
+  const {
+    numStakeWitnesses, numPoolColdWitnesses,
+  } = _countWithdrawalAndCertificateWitnesses(params.unsignedTxParsed)
+  const numMintWitnesses = params.unsignedTxParsed.mint?.length || 0
+
+  if (stakeSigningFiles.length > numStakeWitnesses) {
+    throw Error(Errors.TooManyStakeSigningFilesError)
+  }
+  if (poolColdSigningFiles.length > numPoolColdWitnesses) {
+    throw Error(Errors.TooManyPoolColdSigningFilesError)
+  }
+  if (mintSigningFiles.length > numMintWitnesses) {
+    throw Error(Errors.TooManyMintSigningFilesError)
+  }
+  if (multisigSigningFiles.length > 0) {
+    throw Error(Errors.TooManyMultisigSigningFilesError)
+  }
+}
+
+const validateOrdinarySigning = (params: SigningParameters) => {
+  // these checks should be performed only with createSignedTx call (when the signing file set is
+  // expected to be complete) on top of validateOrdinaryWitnesses
+  const {
+    paymentSigningFiles, stakeSigningFiles, poolColdSigningFiles, mintSigningFiles,
+  } = filterSigningFiles(params.hwSigningFileData)
+
+  if (paymentSigningFiles.length === 0) {
+    throw Error(Errors.MissingPaymentSigningFileError)
+  }
+
+  const {
+    numStakeWitnesses, numPoolColdWitnesses,
+  } = _countWithdrawalAndCertificateWitnesses(params.unsignedTxParsed)
+  const numMintWitnesses = params.unsignedTxParsed.mint?.length || 0
+
+  if (numStakeWitnesses > 0 && (stakeSigningFiles.length === 0)) {
+    throw Error(Errors.MissingStakeSigningFileError)
+  }
+  if (numPoolColdWitnesses > 0 && (poolColdSigningFiles.length === 0)) {
+    throw Error(Errors.MissingPoolColdSigningFileError)
+  }
+  if (numMintWitnesses > 0 && (mintSigningFiles.length === 0)) {
+    throw Error(Errors.TooManyMintSigningFilesError)
+  }
+}
+
+const validatePoolRegistrationTx = (unsignedTxParsed: _UnsignedTxParsed) => {
+  _validateNonMultisigTx(unsignedTxParsed)
+
+  if (unsignedTxParsed.inputs.length === 0) {
```

Comment:
Why are we checking this here? AFAIK having at least one input is mandatory for any tx (to avoid tx replays, esp. because of certificates).

Reply:
Moved to `validateWitnessing` and `validateSigning` - https://github.com/vacuumlabs/cardano-hw-cli/pull/108/commits/35fc2a62f021a63ab9731ab5cb039fdf34b76bc8.